test: Add multi node consolidation metrics

clbar-aws opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.
-->

**Description**

This PR adds three new Prometheus metrics to monitor the efficiency and performance of the binary search algorithm used in multi-node consolidation. These metrics provide visibility into:

- How many iterations the binary search takes to find optimal consolidation batches
- The actual size of consolidation batches achieved
- Time wasted on failed consolidation attempts

**How was this change tested?**

Validated using KWOK provider by scaling a deployment from 50→5 replicas, which triggered multi-node consolidation that successfully consolidated 2 nodes. All three metrics correctly reported values showing 1 iteration per batch, average batch size of 2 nodes, and zero failed simulations.

## Example Output

```javascript
# HELP karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations Alpha metric prone to change. Number of binary search iterations to find optimal multi-node consolidation batch.
# TYPE karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations histogram
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="1"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="2"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="3"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="4"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="5"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="6"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="7"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_bucket{le="+Inf"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_sum 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_iterations_count 2

# HELP karpenter_voluntary_disruption_multi_node_consolidation_batch_size Alpha metric prone to change. Number of nodes in successful multi-node consolidation batch, labeled with decision type.
# TYPE karpenter_voluntary_disruption_multi_node_consolidation_batch_size histogram
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="2"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="3"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="4"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="5"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="7"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="10"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="15"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="20"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="25"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="30"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="40"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="50"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_bucket{decision="delete",le="+Inf"} 2
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_sum{decision="delete"} 4
karpenter_voluntary_disruption_multi_node_consolidation_batch_size_count{decision="delete"} 2

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Clabes committed:
Add metrics to track multi-node consolidation binary search performance

This adds three new Prometheus metrics to monitor the efficiency of the
binary search algorithm used in multi-node consolidation:

1. karpenter_voluntary_disruption_multi_node_consolidation_iterations
   - Histogram tracking number of binary search iterations needed
   - Helps identify if the algorithm is performing efficiently (O(log n))

2. karpenter_voluntary_disruption_multi_node_consolidation_batch_size
   - Gauge showing actual number of nodes consolidated together
   - Labeled by decision type (delete vs replace)
   - Helps understand consolidation effectiveness

3. karpenter_voluntary_disruption_multi_node_consolidation_failed_iteration_duration_seconds
   - Histogram tracking cumulative time spent on failed iterations
   - Identifies computational waste in the binary search process

These metrics provide visibility into the multi-node consolidation algorithm's
performance and help identify optimization opportunities.

Clabes committed:
Optimize iteration histogram buckets to 1-7

Binary search on max 100 candidates has at most log₂(100) ≈ 7 iterations,
so removed unnecessary buckets beyond 7.

jamesmt-aws commented on pkg/controllers/disruption/multinodeconsolidation.go:
```diff
@@ -124,10 +124,15 @@ func (m *MultiNodeConsolidation) firstNConsolidationOption(ctx context.Context,
 	lastSavedCommand := Command{}
+	iterationCount := 0
+	var failedIterationTime time.Duration
+
```
Imagine a case where the binary search succeeds after two rounds. So the whole process could be encoded as FFS, for Fail Fail Succeed. The time for F and S are tracked separately under the current design, so FFS increments both timers even though the aggregate operation eventually succeeded. 

So FFS today, there's allocation of time to both the F timer and the S timer. But I could imagine an argument where the metrics aren't per-call, but are per-search. So FFFFFFF would be a fail, but S, FS, FFS, etc would all be successes. What's the right thing to do from an analysis point of view with FFS?

jamesmthompson reviewed (commented):
Thanks for sending this out! I had a couple of questions about how we're thinking about using these.

Clabes committed:
Refine multi-node consolidation metrics

Changes:
1. Convert batch_size from Gauge to Histogram with buckets [2-50]
   - Better tracks distribution of consolidation batch sizes over time
   - Buckets: 2, 3, 4, 5, 7, 10, 15, 20, 25, 30, 40, 50

2. Add sim_success label to failed_iteration_duration metric
   - Tracks whether binary search ultimately succeeded despite failed iterations
   - Values: "true" (found valid consolidation) or "false" (no valid consolidation)
   - Only records when failedIterationTime > 0

Clabes committed:
Add alpha warning to multi-node consolidation metrics

Mark the three new multi-node consolidation metrics as alpha to signal
they are experimental and may change:
- multi_node_consolidation_iterations
- multi_node_consolidation_batch_size
- multi_node_consolidation_failed_iteration_duration_seconds

This gives flexibility to refine buckets, labels, or metric definitions
based on real-world usage patterns.

clbar-aws reviewed (commented):
Changed from gauge to histogram for successful batch size metric. Added indicator of simulation success/failure with the time spent in failed simulations

Clabes committed:
updated metric types and names

Clabes committed:
included bucket of size 100 for Batch Size metric
